### PR TITLE
Allow action completion to be deferred.

### PR DESCRIPTION
**Pull request by jander-msft**
This change allows action completion to be deferred by default such that all actions are startup asynchronously (but in order), signal startup has completed, and then wait for completion on all deferred action completions. Actions can opt into synchronous completion by setting the WaitForCompletion property to true e.g.

```json
{
  "Type": "CollectTrace",
  "Settings": { ... },
  "WaitForCompletion": true
}
```

### CODE REVIEW

File: src/Tools/dotnet-monitor/CollectionRules/CollectionRulePipeline.cs [hunk 1]
```diff
@@ -226,5 +229,15 @@ public new Task StopAsync(CancellationToken token)
         {
             return base.StopAsync(token);
         }
+
+        // Ensures that the start callback is only invoked once.
+        private void InvokeStartCallback()
+        {
+            if (_invokedStartCallback)
```
Comment (kkeirstead) [inline src/Tools/dotnet-monitor/CollectionRules/CollectionRulePipeline.cs:236]: Should this be negated? As in, if we haven't invoked the callback, we do it (and set it to true) so we can only invoke it once?

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/ActionListExecutorTests.cs [hunk 2]
```diff
@@ -75,23 +92,51 @@ public async Task ActionListExecutor_SecondActionFail()
...
+
                 CollectionRuleActionExecutionException actionExecutionException = await Assert.ThrowsAsync<CollectionRuleActionExecutionException>(
-                    () => executor.ExecuteActions(context, cancellationTokenSource.Token));
+                    () => executor.ExecuteActions(context, startCallback, cancellationTokenSource.Token));
 
                 Assert.Equal(1, actionExecutionException.ActionIndex);
 
                 Assert.Equal(string.Format(Strings.ErrorMessage_NonzeroExitCode, "1"), actionExecutionException.Message);
+
+                if (waitForCompletion)
```
Comment (kkeirstead) [inline src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/ActionListExecutorTests.cs:105]: I realize it's really simple, so up to you, but for code reuse should this be put into a method? Just thinking that if we bulk up `ActionListExecutorTests` in the future we're going to have these lines repeated in even more places.

File: src/Tools/dotnet-monitor/CollectionRules/ActionListExecutor.cs [hunk 3]
```diff
@@ -33,47 +35,100 @@ public ActionListExecutor(ILogger<CollectionRuleService> logger, ICollectionRule
...
+                            throw new InvalidOperationException(Strings.ErrorMessage_CouldNotMapToAction);
+                        }
+
+                        ICollectionRuleAction action = factory.Create(context.EndpointInfo, actionOption.Settings);
+
+                        try
+                        {
+                            await action.StartAsync(cancellationToken);
+
+                            // Check if the action completion should be awaited synchronously (in respected to
```
Comment (kkeirstead) [inline src/Tools/dotnet-monitor/CollectionRules/ActionListExecutor.cs:68]:
```suggestion
                            // Check if the action completion should be awaited synchronously (in respect to
```

File: src/Tools/dotnet-monitor/CollectionRules/ActionListExecutor.cs [hunk 4]
```diff
@@ -33,47 +35,100 @@ public ActionListExecutor(ILogger<CollectionRuleService> logger, ICollectionRule
...
+                            // starting the next action). If not, add a deferred entry so that it can be completed
+                            // add the end of the execution list.
```
Comment (kkeirstead) [inline src/Tools/dotnet-monitor/CollectionRules/ActionListExecutor.cs:70]: I'm guessing this should be "at" or "after" the end of the execution list, but I may be misinterpreting the comment

File: src/Tools/dotnet-monitor/CollectionRules/ActionListExecutor.cs [hunk 5]
```diff
@@ -33,47 +35,100 @@ public ActionListExecutor(ILogger<CollectionRuleService> logger, ICollectionRule
...
 
-                    ICollectionRuleAction action = factory.Create(context.EndpointInfo, actionOption.Settings);
+                    ++actionIndex;
+                }
+
+                // Notify that all actions have started
+                startCallback?.Invoke();
 
+                // Wait for any actions whose completion has been deferred.
+                for (int i = 0; i < deferredCompletions.Count; i++)
```
Comment (kkeirstead) [inline src/Tools/dotnet-monitor/CollectionRules/ActionListExecutor.cs:102]: More-so a thought than asking for a change, but I was curious about why we don't use a while loop here (e.g. `while (deferredCompletions.Count > 0`). Then we wouldn't have to play the game with incrementing and decrementing `i`, and when we assign `deferredCompletion` we could always use `deferredCompletions[0]`. In my head at least that feels more intuitive, but this also works.

Comment (wiktork) [reply src/Tools/dotnet-monitor/CollectionRules/ActionListExecutor.cs:102]: nit: Alternatively consider a queue

**Review by wiktork — APPROVED** [review]

File: src/Tools/dotnet-monitor/CollectionRules/CollectionRulePipeline.cs (hunk 1 above)
Comment (jander-msft) [reply src/Tools/dotnet-monitor/CollectionRules/CollectionRulePipeline.cs:236]: Good catch; I'll adjust the tests to make sure this does in fact work after fixing it.

File: src/Tools/dotnet-monitor/CollectionRules/ActionListExecutor.cs (hunk 4 above)
Comment (jander-msft) [reply src/Tools/dotnet-monitor/CollectionRules/ActionListExecutor.cs:70]: Either way, but yet, you are correct; I'll go with "after".

**Pull request merged**